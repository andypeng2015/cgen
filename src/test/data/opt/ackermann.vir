module foo

export function w $foo(w %m, w %n) {
@start:
  %c = eq.w %m, 0_w
  br %c, @n1, @cmgt
@n1:
  %r = add.w %n, 1_w
  ret %r
@cmgt:
  %c = gt.w %m, 0_w
  br %c, @cnzero, @cngt
@cnzero:
  %c = eq.w %n, 0_w
  br %c, @case1, @cngt
@case1:
  %m = sub.w %m, 1_w
  %r = call.w $foo(%m, 1_w)
  ret %r
@cngt:
  %c = gt.w %n, 0_w
  br %c, @case2, @bad
@case2:
  %n = sub.w %n, 1_w
  %r = call.w $foo(%m, %n)
  %m = sub.w %m, 1_w
  %r = call.w $foo(%m, %r)
  ret %r
@bad:
  hlt
}
