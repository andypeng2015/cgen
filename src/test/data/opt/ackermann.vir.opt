module foo

export function w $foo(w %m, w %n) {
@22:
  jmp @2(%m, %n)
@2(%1, %2):
  %3 = eq.w %1, 0x0_w ; @23
  br %3, @3, @4
@3:
  %7 = add.w %2, 0x1_w ; @27
  ret %7
@4:
  %5 = eq.w %2, 0x0_w ; @25
  %4 = sub.w %1, 0x1_w ; @24
  br %5, @2(%4, 0x1_w), @8
@8:
  %6 = sub.w %2, 0x1_w ; @26
  %r.1 = call.w $foo(%1, %6) ; @18
  jmp @2(%4, %r.1)
}
