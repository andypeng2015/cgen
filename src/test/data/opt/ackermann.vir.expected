module foo

export function w $foo(w %m, w %n) {
@20:
  jmp @0(%m, %n)
@0(%1, %2):
  %3 = eq.w %1, 0x0_w ; @21
  br %3, @1, @2
@1:
  %8 = add.w %2, 0x1_w ; @26
  ret %8
@2:
  %4 = eq.w %2, 0x0_w ; @22
  br %4, @8, @6
@8:
  %5 = sub.w %1, 0x1_w ; @23
  jmp @0(%5, 0x1_w)
@6:
  %6 = sub.w %2, 0x1_w ; @24
  %r.1 = call.w $foo(%1, %6) ; @16
  %7 = sub.w %1, 0x1_w ; @25
  jmp @0(%7, %r.1)
}
